### PR TITLE
Symink Process & Exception Handling

### PR DESCRIPTION
**Pull request by mikeymike**
Creating this for viewing/reviewing purposes, please don't merge 👍 😄 
- Refactor IO
- Install still NullIO by default
- IoInterface is ConsoleIO by default
- IO all respects verbosity levels
- Remove link/unlink commands
- Add Linker class to handle links
- Detect whether to use system for binary
- Still very WIP.

### CODE REVIEW

**Comment by AydinHassan** [comment]
LGTM! What's left?

**Comment by mikeymike** [comment]
I want to go over it a bit more and do a general cleanup (did most of this if not all of it last night though) ... then the only other bit is I want to do unix detection in the app bootstrap. If someones running in Windows but not Cygwin or some other unix based env we should fail nicely with some instructions on setting up Cygwin or something. At the minute it will just run and fail miserably with paths 😂 

Also need to test this shit in Windows 👍

File: app/bootstrap.php [hunk 1]
```diff
@@ -27,5 +27,17 @@
     ->useAutowiring(false)
     ->build();
 
+if (!file_exists('/usr/local/bin')) {
```
Comment (mikeymike) [inline app/bootstrap.php]: @AydinHassan I'm pretty happy with everything except this check. Wanted a simple as possible check to see if we're running on a Unix based system. Something tells me it's not a safe check really.

I don't really want to check for Cygwin because it looks like there is also a whole bunch of other Cywin like apps for Windows which could just get messy. 

Maybe `PHP_OS` to see if it's Windows as with Cygwin `PHP_OS` is  literally just `Cygwin` so I guess other similar apps will do the same

Comment (AydinHassan) [inline app/bootstrap.php]: what do you need to do different on Windows/Linux? What is the check for?

Comment (AydinHassan) [inline app/bootstrap.php]: ah I see - Can't you just check the directory separator, like Symfony does?

Comment (mikeymike) [inline app/bootstrap.php]: good call... I'll check to make sure the results are right in Cygwin, if they are I'll go with that approach.

**Comment by mikeymike** [comment]
@AydinHassan I think I'm done with this bro.. can you give it a once over and let me know what you think ❤️

File: app/bootstrap.php [hunk 2]
```diff
@@ -27,5 +27,17 @@
     ->useAutowiring(false)
     ->build();
 
+if (DIRECTORY_SEPARATOR === '\\') {
+    $container->get(\Composer\IO\IOInterface::class)->write([
+        '',
+        ' Woops!... It looks like your not running in a Unix envirnoment',
+        '',
+        ' Currently we only support Unix based systems, if your running Windows please use Cygwin',
```
Comment (AydinHassan) [inline app/bootstrap.php]: `your` should be `you're`

File: src/Command/UninstallCommand.php [hunk 3]
```diff
@@ -53,19 +68,30 @@ protected function configure()
...
+        try {
+            $this->linker->unlink($workshop, $input->getOption('force'));
             $this->uninstaller->uninstallWorkshop($workshop);
         } catch (WorkshopNotInstalledException $e) {
-            $output->writeln(sprintf(' <error>Workshop "%s" not currently installed</error>', $workshop));
+            $output->writeln(sprintf(' <error> Workshop "%s" not currently installed </error>', $workshop->getName()));
             return;
         } catch (\RuntimeException $e) {
-            $output->writeln(sprintf(' <error>Failed to uninstall workshop "%s"</error>', $workshop));
+            $output->writeln([
```
Comment (AydinHassan) [inline src/Command/UninstallCommand.php:89]: Maybe we should give a bit more info on failures? Passing the exception message?

Comment (AydinHassan) [inline src/Command/UninstallCommand.php:89]: Actually - if we are in verbose mode, i'd print the whole stack.

File: src/Installer.php [hunk 4]
```diff
@@ -38,31 +38,38 @@
...
      * @param Filesystem $filesystem
      * @param Factory $factory
+     * @param IOInterface $io
      */
     public function __construct(
         ManagerState $state,
         Downloader $downloader,
         Filesystem $filesystem,
-        Factory $factory
+        Factory $factory,
```
Comment (AydinHassan) [inline src/Installer.php]: Maybe alias this to `ComposerFactory` or sumthin

File: src/Installer.php [hunk 5]
```diff
@@ -38,31 +38,38 @@
...
      * @param Workshop $workshop
-     * @param IOInterface $io
      * @throws \League\Flysystem\FileExistsException
      */
-    public function installWorkshop(Workshop $workshop, IOInterface $io)
+    public function installWorkshop(Workshop $workshop)
     {
-        if ($this->state->isWorkshopInstalled($workshop->getName())) {
+        if ($this->state->isWorkshopInstalled($workshop)) {
             throw new WorkshopAlreadyInstalledException;
```
Comment (AydinHassan) [inline src/Installer.php:79]: Maybe pass the workshop in to this exception so a better error can be printed: `WorkshopAlreadyInstalledException::fromWorkshop($workshop);` ?

File: src/Linker.php [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+    /**
+     * @var Filesystem
+     */
+    private $filesystem;
+
+    /**
+     * @var bool
+     */
+    private $useSystem;
```
Comment (AydinHassan) [inline src/Linker.php]: What is this about?

File: src/Linker.php [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        $this->removeWorkshopBin($localTarget, $force);
+
+        return $this->useSystem
+            ? $this->link($workshop, $localTarget) && $this->symlinkToSystem($workshop, $force)
+            : $this->link($workshop, $localTarget);
+    }
+
+    /**
+     * @param Workshop $workshop
+     * @param $force
```
Comment (AydinHassan) [inline src/Linker.php]: `bool`

File: src/Linker.php [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        }
+
+        $this->removeWorkshopBin($systemTarget, $force);
+
+        return $this->link($workshop, $systemTarget);
+    }
+
+    /**
+     * @param Workshop $workshop
+     * @param $target
```
Comment (AydinHassan) [inline src/Linker.php]: `string`

File: src/Linker.php [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+     *
+     * @return bool
+     */
+    private function link(Workshop $workshop, $target)
+    {
+        if (!symlink($this->getWorkshopSrcPath($workshop), $target)) {
```
Comment (AydinHassan) [inline src/Linker.php]: you maybe need to silence this, I think it produces warnings on failure.

File: src/Linker.php [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+    /**
+     * @param Workshop $workshop
+     * @param bool $force
+     *
+     * @return bool
+     */
+    public function symlink(Workshop $workshop, $force = false)
+    {
+        if (!$this->state->isWorkshopInstalled($workshop)) {
```
Comment (AydinHassan) [inline src/Linker.php:51]: `unlink` throws an exception - any reason why `symlink` returns false?

File: src/Linker.php [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+        return $this->filesystem->getAdapter()->applyPathPrefix($path);
+    }
+
+    /**
+     * @param string $binary
+     * @return string
+     */
+    private function getSystemInstallPath($binary)
+    {
+        return sprintf('/usr/local/bin/%s', $binary);
```
Comment (AydinHassan) [inline src/Linker.php:226]: Does this exist on cygwin?

File: src/Linker.php (hunk 6 above)
Comment (mikeymike) [inline src/Linker.php]: Not sure why I put it as a property tbh ... but it's just a flag to state whether or not to symlink into the system path as well as the phpschool dir in the users home. 

the idea being that if the phpschool dir is in the users path we don't need to symlink to the system dir.

File: src/Linker.php (hunk 9 above)
Comment (mikeymike) [inline src/Linker.php]: [docs](http://php.net/manual/en/function.symlink.php) doesn't have any warnings ... just bool return

File: src/Linker.php (hunk 6 above)
Comment (AydinHassan) [inline src/Linker.php]: Ok cool makes sense, maybe add some text somewhere describing the process of checking the dirs :)

File: src/Linker.php (hunk 9 above)
Comment (AydinHassan) [inline src/Linker.php]: it definitely produces notices/warning - trust me :D

File: src/Linker.php (hunk 6 above)
Comment (mikeymike) [inline src/Linker.php]: tbh I think I should move it to a method instead and can document in the doc block  👍

File: src/Linker.php (hunk 9 above)
Comment (mikeymike) [inline src/Linker.php]: 😂 oki doke, better to be safe than sorry I guess

File: src/Linker.php (hunk 11 above)
Comment (mikeymike) [inline src/Linker.php:226]: from my testing it does 👍 ... well it installs the things so it must do 😂

File: src/Installer.php (hunk 5 above)
Comment (mikeymike) [inline src/Installer.php:79]: ahh missed this... will add asap makes sense rather than relying on the caller/exception handler to already have the workshop

**Comment by mikeymike** [comment]
Updated. Not had chance to test in Windows and I'm starting to write some tests now to test all the edge cases exception handling etc. 

with that said though the tests will probably take a while 😨 😂
